File: tests/bucket_test.rs
mod common;
use objectscale_client::bucket::{BucketBuilder, BucketTag};
use objectscale_client::{iam::AccountBuilder, tenant::TenantBuilder};

#[test]
fn test_bucket() {
    let mut management_client = common::create_management_client();
    let mut objectstore_client = common::create_objectstore_client();

    let account_name = "testbucket";
    let account = AccountBuilder::default()
        .alias(account_name)
        .encryption_enabled(false)
        .build()
        .expect("build account");
    let account = management_client
        .create_account(account)
        .expect("create account");

    let tenant_alias = "testbucket";
    let tenant = TenantBuilder::default()
        .alias(tenant_alias)
        .id(&account.account_id)
        .is_encryption_enabled(false)
        .build()
        .expect("build tenant");
    let _ = objectstore_client
        .create_tenant(tenant)
        .expect("create tenant");

    let bucket_name = "testbucket";
    let expiration = -1;
    let bucket = BucketBuilder::default()
        .name(bucket_name)
        .namespace(&account.account_id)
        .audit_delete_expiration(expiration)
        .tags(vec![BucketTag {
            key: "key1".to_string(),
            value: "value1".to_string(),
        }])
        .build()
        .expect("build bucket");
    let _ = objectstore_client
        .create_bucket(bucket)
        .expect("create bucket");

    let mut bucket = objectstore_client
        .get_bucket(&bucket_name, &account.account_id)
        .expect("get bucket");
    assert_eq!(bucket.name, bucket_name);
    assert_eq!(bucket.namespace, account.account_id);
    assert_eq!(bucket.audit_delete_expiration, expiration);
    assert_eq!(bucket.tags.len(), 1);

    let new_expiration = 0;
    bucket.audit_delete_expiration = new_expiration;
    let bucket = objectstore_client
        .update_bucket(bucket)
        .expect("update bucket");
    assert_eq!(bucket.audit_delete_expiration, new_expiration);
    assert_eq!(bucket.name, bucket_name);
    assert_eq!(bucket.namespace, account.account_id);
    assert_eq!(bucket.tags.len(), 1);

    let buckets = objectstore_client
        .list_buckets(&account.account_id, "")
        .expect("list buckets");
    assert_ne!(buckets.len(), 0);

    objectstore_client
        .delete_bucket(&bucket_name, &account.account_id, false)
        .expect("delete bucket");

    objectstore_client
        .delete_tenant(&account.account_id)
        .expect("delete tenant");
    management_client
        .delete_account(&account.account_id)
        .expect("delete account");
}

File: golang/tests/bucket_test.go
package tests

import (
	"testing"

	"github.com/stretchr/testify/assert"
	objectscale "github.com/vangork/objectscale-client/golang/pkg"
)

func TestBucket(t *testing.T) {
	manamgentclient := CreateManagementClient(t)
	defer manamgentclient.Close()

	objectstoreClient := CreateObjectstoreClient(t)
	defer objectstoreClient.Close()

	accountName := "testbucket"
	account := &objectscale.Account{
		Alias:             accountName,
		Description:       accountName,
		EncryptionEnabled: false,
	}
	account, err := manamgentclient.CreateAccount(account)
	assert.Nil(t, err)

	alias := "testbucket"
	encryptionEnabled := false
	tenant := &objectscale.Tenant{
		Alias:               alias,
		Id:                  account.AccountId,
		IsEncryptionEnabled: encryptionEnabled,
	}
	_, err = objectstoreClient.CreateTenant(tenant)
	assert.Nil(t, err)

	bucketName := "testbucket"
	expiration := int64(-1)
	bucket := &objectscale.Bucket{
		Name:                  bucketName,
		Namespace:             account.AccountId,
		AuditDeleteExpiration: expiration,
		Tags:                  []objectscale.BucketTag{{Key: "key1", Value: "value1"}, {Key: "key2", Value: "value2"}},
	}
	_, err = objectstoreClient.CreateBucket(bucket)
	assert.Nil(t, err)

	bucket, err = objectstoreClient.GetBucket(bucketName, account.AccountId)
	assert.Nil(t, err)
	assert.Equal(t, bucketName, bucket.Name)
	assert.Equal(t, account.AccountId, bucket.Namespace)
	assert.Equal(t, 2, len(bucket.Tags))
	assert.Equal(t, expiration, bucket.AuditDeleteExpiration)

	newExpiration := int64(0)
	bucket.AuditDeleteExpiration = newExpiration
	bucket, err = objectstoreClient.UpdateBucket(bucket)
	assert.Nil(t, err)
	assert.Equal(t, newExpiration, bucket.AuditDeleteExpiration)
	assert.Equal(t, bucketName, bucket.Name)
	assert.Equal(t, account.AccountId, bucket.Namespace)
	assert.Equal(t, 2, len(bucket.Tags))

	buckets, err := objectstoreClient.ListBuckets(account.AccountId, "")
	assert.Nil(t, err)
	assert.Less(t, 0, len(buckets))

	err = objectstoreClient.DeleteBucket(bucketName, account.AccountId, false)
	assert.Nil(t, err)

	err = objectstoreClient.DeleteTenant(account.AccountId)
	assert.Nil(t, err)

	err = manamgentclient.DeleteAccount(account.AccountId)
	assert.Nil(t, err)
}
